docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,224 @@
+
+```
+$ oras manifest fetch --oci-layout /tmp/ginkgo1163328512
+Error: "/tmp/ginkgo1163328512": no tag or digest when expecting <name:tag|name@digest>
+```
+
+Suggested error message:
+
+```
+$ oras manifest fetch --oci-layout /tmp/ginkgo1163328512
+Error: "/tmp/ginkgo1163328512": no tag or digest specified
+Usage: oras manifest fetch [flags] <name>{:<tag>|@<digest>}
+You need to specify an artifact reference in the form of "<name>:<tag>" or "<name>@<digest>". Run "oras manifest fetch -h" for more options and examples 
+```
+
+#### Example 4: push a manifest if no media type flag provided
+
+Current behavior and output:
+
+```
+$ oras manifest push --oci-layout /sample/images:foobar:mediatype
+Error: media type is not recognized. 
+```
+
+Suggested error message:
+
+```
+$ oras manifest push --oci-layout /sample/images:foobar:mediatype
+Error: media type is not recognized. 
+Usage: oras manifest push [flags] <name>[:<tag>[,<tag>][...]|@<digest>] <file>
+You need to specify a valid media type "mediaType" in the manifest JSON or via the "--media-type" flag.
+```
+
+#### Example 5: attach an artifact if the given option is unknown
+
+Current behavior and output:
+
+```
+$ oras attach --artifact-type oras/test localhost:5000/command/images:foobar --distribution-spec v1.0
+Error: unknown distribution specification flag: v1.0
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type oras/test localhost:5000/sample/images:foobar --distribution-spec ???
+Error: unknown distribution specification flag: "v1.0". 
+Available options: v1.1-referrers-api, v1.1-referrers-tag
+```
+
+#### Example 6: when attaching an file, if no file reference or annotation is provided
+
+Current behavior and output:
+
+```
+$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
+Error: no blob is provided
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
+Error: no file or annotation provided in the command
+Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<type>] [...]
+To attach to an existing artifact, please provide files via argument or annotations via flag "--annotation". Run "oras attach -h" for more options and examples
+```
+
+#### Example 7: When pushing files, if the annotation file doesn't match the required format
+
+Current behavior and output:
+
+```
+$ oras push --annotation-file sbom.json ghcr.io/library/alpine:3.9
+Error: failed to load annotations from sbom.json: json: cannot unmarshal string into Go value of type map[string]map[string]string. Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation-file annotation.json ghcr.io/library/alpine:3.9
+Error response from registry: failed to load annotations from sbom.json: json: cannot unmarshal string into Go value of type map[string]map[string]string. 
```

Comment:
This is not a registry error, suggest to change the error description to:
```suggestion
Error: invalid manifest json file: failed to load annotations from sbom.json
```

Reply:
I remember `json: cannot unmarshal string into Go value of type map[string]map[string]string` is a registry error message. Can we remove it?

Reply:
In this case it's not a registry error. Here annotation file is a user input and JSOM unmarshal goes wrong based on the user input。